### ISSUE TITLE
Preserve logits on device for sampler (vLLM)

### ISSUE THREAD

**Issue by nvukobratTT** [tenstorrent/tt-xla#5069]
### Summary
Today, running LLM models with vLLM introduces separate sampling graphs alongside the main backbone models. The limitation of this is that backbone logits are pulled back to host DRAM, followed by a second graph (sampler), which reads data from the same host DRAM. In sum, a much better solution would be to support that logic stays on-device: DRAM + sampler of choice reading logits from the same device DRAM instead of the host. 

One big performance impact for vLLM models is that running sampling requires that the logic from the backbone model is on the host. 

### Repro
```py
import vllm

def test_llama3_3b_generation():
    prompts = [
        "I like taking walks in the",
    ]
    sampling_params = vllm.SamplingParams(temperature=0.8, top_p=0.95, max_tokens=32)
    llm_args = {
        "model": "meta-llama/Llama-3.2-3B",
        "max_num_batched_tokens": 32,
        "max_num_seqs": 1,
        "max_model_len": 32,
        "gpu_memory_utilization": 0.002,
        "additional_config": {
            "min_context_len": 32,
            "cpu_sampling": False,
            "enable_trace": True,
            "num_hidden_layers": 1,
        },
    }
    llm = vllm.LLM(**llm_args)

    output_text = llm.generate(prompts, sampling_params)[0].outputs[0].text
    print(f"prompt: {prompts[0]}, output: {output_text}")

if __name__ == "__main__":
    test_llama3_3b_generation()
```

**Comment by kmabeeTT** [comment]
Thanks for flagging @nvukobratTT - was surprised to see this since I definitely did some checks a while back to confirm they never left device, but now I see it's an artifact of `enable_trace=True` which was added later.  

Good news is this is already solved 2 days ago with https://github.com/tenstorrent/tt-xla/pull/4789 from @alinakhanTT  that merged backbone and sampler graphs (reduce from 5 to 2 graphs), and another PR will reduce from 2->1 graphs.  Had claude take a look and run some comparisons:


> Reproduced and traced through the TTNN graphs (`enable_trace=True`, `cpu_sampling=False`). Findings:
> 
> **Pre-#4789 (separate graphs):** confirmed the round-trip. The separate sampler graph takes its logits input from **host** memory and re-stages it to DRAM before trace replay:
> ```mlir
> // sampler graph @main
> %arg0: tensor<1x128256xbf16, #system_memory>   // logits arrives from HOST
> "ttnn.write_tensor"(%arg0, %dram_buf)          // H2D copy before trace replay
> ```
> `compute_logits` reads its hidden state from `#system_memory` too, so each decode step round-trips: `backbone → D2H → H2D → compute_logits → D2H → H2D → sampler`.
> 
> **It's trace-specific, not a multi-graph problem.** With `enable_trace=False`, every sampler graph reads logits straight from `#dram` (no `write_tensor`); `system_memory` only appears for the genuine first input into the backbone. Without trace, PJRT keeps intermediates on device across the separate executables; with trace, replay forces host staging of each dispatch's inputs.
> 
> **Already addressed by #4789 (verified).** Cherry-picking the merged 5→2 fusion and re-running: `1x128256` no longer appears as any graph's `#system_memory` arg — logits are now an internal SSA value inside the fused `decode_postprocess`, produced by the LM head and consumed by the sampler in one program:
> ```mlir
> %12 = "ttnn.matmul"(%hidden, %lm_head) ... -> tensor<1x128256xbf16, #dram>   // logits produced on-device
> %17 = "ttnn.argmax"(%16)  : (tensor<1x128256xbf16, ...>) -> tensor<1x1xui32>  // sampled in-place, same graph
> ```
> The vocab-sized per-token round-trip is gone. The only host staging left in trace mode is the small `1x3072` hidden state between backbone and `decode_postprocess`, which **#4833** removes by fusing those two dispatches into one.
> 
> 

Could be marked closed if you're happy with this :)

**Comment by nvukobratTT** [comment]
Nice one! Thanks for letting me know! Closing this one :))
